packages/storage-events/src/move-todo.js

```diff
@@ -0,0 +1,99 @@
+module.exports = { moveTodo };
+
+function moveTodo(todos, event) {
+  const oldNode = findTodo(todos, event.data.todoId);
+  const position = event.data.position;
+  const newParentId = event.data.parentId;
+  const newParentIsChild = isIdInTodos([oldNode], newParentId);
+  if (newParentIsChild) {
+    return todos;
+  }
+
+  todos = detachFromParent(todos, oldNode, newParentId);
+  todos = attachToNewParent(todos, oldNode, position, newParentId);
+
+  return todos;
+}
+
+function detachFromParent(todos, oldNode, newParentId) {
+  // has it been a root node?
+  if (oldNode.parentId === null || oldNode.parentId === undefined) {
+    todos = removeFromBranch(todos, oldNode.todoId);
+  } else {
+    const oldParent = findTodo(todos, oldNode.parentId);
+    oldParent.children = removeFromBranch(oldParent.children, oldNode.todoId);
+    todos = replaceNode(todos, oldParent.todoId, oldParent);
+  }
+  return todos;
+}
+
+function attachToNewParent(todos, oldNode, position, newParentId) {
+  // update the node values to reflect its new place
+  oldNode.position = position;
+  oldNode.parentId = newParentId;
+  // is it gonna be a root node?
+  if (newParentId === null) {
+    todos = insertIntoBranch(todos, oldNode, position);
+  } else {
+    const newParent = findTodo(todos, newParentId);
+    newParent.children = insertIntoBranch(newParent.children, oldNode, position);
+    todos = replaceNode(todos, newParent);
+  }
+  return todos;
+}
+
+function removeFromBranch(branch, todoId) {
+  branch.sort(positionSortFunction);
+  let result = branch.filter(x => x.todoId != todoId);
+  // adjust position values
+  for (let i = 0; i < result.length; i++) {
+    result[i].position = i;
+  }
+  return result.sort(positionSortFunction);
```

Comment:
Why does it have to be sorted again, when you set the position to their indices?

Reply:
Unnecessary, will remove